[pipeline-ui] Retrieve pod logs from argo archive

eterna2 opened the pull request:
A possible workaround for #1803 

Currently, I persist pod logs with argo's archiveLogs config to s3 bucket. 

This PR allow the UI to fallback to retrieving GCed pods logs from the argo archive artifactory - this can be either minio store or a s3 bucket.

This feature can be enabled by setting the following env variables:
```js
  /** Is Argo log archive enabled? */
  ARGO_ARCHIVE_LOGS = "true",
  /** Use minio or s3 client to retrieve archives. */
  ARGO_ARCHIVE_ARTIFACTORY = 'minio',
  /** Bucket to retrive logs from */
  ARGO_ARCHIVE_BUCKETNAME = 'mlpipeline',
  /** Prefix to logs. */
  ARGO_ARCHIVE_PREFIX = 'logs',
```

### Updates 20 sep 2019:
Investigate the feasibility of getting pod logs archive location from argo workflow status:
- conclusion is yes, but if the pipeline did not complete successfully, the status will not be updated with the output information (even if the logs are actually archived)

##### Changes
- Updated @types/minio  to match the minio package (`useSSL` instead of `insecure` <- old)
- Created a few additional helpers modules:
  - `aws-helper` provides utils to query and handles AWS instance profile session credentials (aka `kube2iam`  or ec2 profiles can be used instead of providing access key and secret to minio client)
  - `workflow-helper` provides utils to retrieve pod logs archive info from argo workflow status
  - `minio-helper` provides utils to retrieve objects from minio or s3 backends, includes capability to use AWS ec2 credentials to access s3
- Replace existing handlers for `s3` and `minio` objects with utils from `minio-helper`
- Update get pod logs handler to try getting the pod logs in the following orders:
  - get logs from pod with k8s api
  - get archive location from workflow status, and retrieve from archive
  - if workflow status does not have required info, fallback to the values provided in the environment variables (by user) and try to retrieve the logs.

- I have build the image and test on my own cluster in AWS. However, i only tested on the pod logs, have not really test on the ui-metadata yet.

### TODO
- probably shld use the schema from third-party/argo
- some unit tests?

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/kubeflow/pipelines/2081)
<!-- Reviewable:end -->

Ark-kun commented on frontend/server/server.ts:
```diff
@@ -88,6 +96,13 @@ const s3Client = new MinioClient({
+/** helper function to retrieve pod logs from argo artifactory. */
+const getPodLogsFromArtifactory = _as_bool(ARGO_ARCHIVE_LOGS) ? k8sHelper.getPodLogsFromArtifactoryHelper(
+  ARGO_ARCHIVE_ARTIFACTORY==='minio' ? minioClient : s3Client,
+  ARGO_ARCHIVE_BUCKETNAME,
```
Can the actual log location URL be taken from the Workflow status object?

Ark-kun replied:
> Just a question on which client to use.

AFAIK, in the Frontend the `s3Client` is actually and instance of `minioClient`. There is some weird logic where `s3` reads data uncompressed while `minio` expects `tar.gz`.

Ark-kun replied:
Generally, I feel that the UI handling of artifacts can get some improvement. If you see some ways to streamline and improve that part, I hope @jingzhang36 and other people working on UX will be glad to see that.

Ark-kun replied:
@eterna2 Do you think you can take on this: https://github.com/kubeflow/pipelines/issues/2172 later ?

eterna2 replied:
Sure.

Ark-kun commented:
/ok-to-test

Ark-kun commented:
/retest

Ark-kun commented:
/retest

eterna2 commented:
@Ark-kun 
Made a big change to the code base.
Basically, I added a few new features:

### pod logs
workflow status does provide the archive information only if the job completes without errors. So we might still need the user-provided env variables as an optional fallback.

So the logic now is:
- query k8s for logs
- if fails, query for workflow, and try extract log location, create required client, and get the logs
- else, try to retrieve based on the info provided in the env
- if all else fails, return error message

### AWS instance profile
A wrapper over the minio-js to mimic similar behavior as minio-go, so that pipeline ui can retrieve objects from s3 w/o access key/secrets, and instead use the ec2 instance profile (or kube2iam credentials). 

This means if the IAM role for the pod/ec2 running the pipeline-ui can be used to interact with s3 instead of access key/secret.

### k8s role/service account
Updated manifest for ui role to be able to get secrets as well as workflow.

Ark-kun commented on frontend/server/server.ts:
```diff
@@ -199,48 +221,26 @@ const artifactsHandler = async (req, res) => {
-      });
+    case 'minio':
+      try {
+        res.send(await getTarObjectAsString({bucket, key, client: minioClient}));
```
In the next PR, can you please make a function that auto-detects whether the data is gzipped and unpacks if needed?

eterna2 replied:
I am thinking whether to change from `node-tar` to `tar-stream` + `maybe-gzip`.

But probably need more testing to ensure it works.

Ark-kun commented:
I think this is a good PR that also improves the handling of data access by Frontend.
/lgtm
/cc @Bobgy @jingzhang36 - Can you please review the Frontend code.

Bobgy commented:
@eterna2 Thanks for the contribution! This is a great PR.

I briefly went through the change, and there's no obvious problems. I think you can start on some testing.

Our integration test infra doesn't run tests on AWS, so you cannot rely on that. I recommend unit tests for the node server that covers end to end flow of your use case. Please keep in mind we can only rely on your tests to make sure we don't break your feature, so prefer tests that cover the whole flow, rather than implementation details.

I need some time to gather enough context about minio, argo workflow and the exact approach you took.  I will add more comments when I have time for a more detailed review.

Ark-kun commented:
@Bobgy Let's try to prioritize this PR a bit since there are some UX improvements that might depend on some of the refactorings introduced here.

Bobgy commented on manifests/kustomize/base/pipeline/ml-pipeline-ui-role.yaml:
```diff
@@ -14,6 +14,13 @@ rules:
   - create
   - get
   - list
+- apiGroups:
```
you should reflect the same changes here too: https://github.com/kubeflow/pipelines/blob/master/manifests/gcp_marketplace/chart/kubeflow-pipelines/templates/pipeline.yaml#L148

eterna2 replied:
rebased and updated the helm chart.

eterna2 commented:
@Bobgy

I need some time to write the tests. I dun see anything specific currently in this repo thou. I do see some tests far down the line, but those are more end-2-end tests.

How do u want me to include the tests? Would adding `tests/` folder and update package json and update the build script works?

Standard jest tests that mocks some of the AWS and K8s requests to test the various fallbacks.

@ark-kun
Might need some time for the ux artifacts fix. Cuz I am on vacation. Not allowed to touch my computer until I am back next mon.

eterna2 commented on frontend/server/aws-helper.ts:
```diff
@@ -0,0 +1,88 @@
+    /**
+     * Get the AWS metadata store session credentials.
+     */
+    async get_credentials(): Promise<IAWSMetadataCredentials> {
```
fixed

eterna2 commented on frontend/server/server.ts:
```diff
@@ -70,24 +79,37 @@ const MINIO_ENDPOINT = MINIO_NAMESPACE && MINIO_NAMESPACE.length > 0 ? `${MINIO_
 
 /** minio client for minio storage */
-const minioClient = new MinioClient({
+const minioOptions: MinioClienOptions = {
```
fixed

eterna2 commented on frontend/server/minio-helper.ts:
```diff
@@ -0,0 +1,66 @@
+ * Create minio client with aws instance profile credentials if needed.
+ * @param config minio client options where `accessKey` and `secretKey` are optional.
+ */
+export async function createMinioClient(config: Partial<MinioClientOptions>) {
```
fixed

Bobgy commented on frontend/server/minio-helper.ts:
```diff
@@ -24,15 +24,20 @@ export interface IMinioRequestConfig {
+            const credentials = await awsInstanceProfileCredentials.getCredentials();
             if (credentials) {
               const {AccessKeyId: accessKey, SecretAccessKey: secretKey, Token: sessionToken} = credentials;
               return new MinioClient({...config, accessKey, secretKey, sessionToken} as MinioClientOptions);
```
nit: Can you remove the type cast here? with your change, it should work.

eterna2 replied:
fixed

Bobgy commented:
> @Bobgy
> 
> I need some time to write the tests. I dun see anything specific currently in this repo thou. I do see some tests far down the line, but those are more end-2-end tests.

Sorry I didn't realize there are no existing tests that you can easily add your tests to. I changed my mind, as long as you tested your solution, I don't think this is blocking.

I will be very grateful if you want to contribute tests in a separate PR. I can also at least setup test infra for tests I mentioned. I don't think that can be on my priority, so let me know if you would like to take this.

> How do u want me to include the tests? Would adding `tests/` folder and update package json and update the build script works?

For tests that covers a lot of features, you can do that in `frontend/server/tests` folder. For unit tests that cover just one file, do it in the same place as the file like `module.test.ts`.

> Standard jest tests that mocks some of the AWS and K8s requests to test the various fallbacks.

Yes, standard jest tests that mock some external requests would be the best. No need to pursue coverage at the moment.

> @Ark-kun
> Might need some time for the ux artifacts fix. Cuz I am on vacation. Not allowed to touch my computer until I am back next mon.

Can you share the issue? I don't have much context with this.

Bobgy commented:
Created issue for frontend node server test infra: https://github.com/kubeflow/pipelines/issues/2215

eterna2 commented:
> Sorry I didn't realize there are no existing tests that you can easily add your tests to. I changed my mind, as long as you tested your solution manual, I don't think adding tests is a blocking issue.

Cool. I have tested on my own cluster and it works fine. 

> I will be very grateful if you want to contribute tests in a separate PR. I can also at least setup test infra for tests I mentioned. I don't think that can be on my priority, so let me know if you would like to take this.

Yup. I can make a PR for the tests. But probably only after next couple of weeks. But let's keep it flexible, cuz I can't confirm my availability yet.

> Can you share the issue? I don't have much context with this.
#2172

Btw can u get someone to have a look at #2080 ? It is the last PR I need to totally migrate my kfp deployment to S3 directly (remove the need for me to deploy minio to proxy s3).

eterna2 commented:
@Bobgy 
Made a small bug fix. Just discovered I forgot to make s3client a callable instead of a promise to a client.

The reason is because AWS instance profile will expire after X minutes. I need to recreate the client if the token has expired. So instead of a client, it shld be a function which returns a client.

eterna2 commented on frontend/server/server.ts:
```diff
@@ -70,24 +79,37 @@ const MINIO_ENDPOINT = MINIO_NAMESPACE && MINIO_NAMESPACE.length > 0 ? `${MINIO_
-} as any);
+};
+/** minio client to s3 with AWS instance profile IAM if access keys are not provided. */
+const s3ClientPromise = () => createMinioClient(s3Options);
```
changed this to a callable to `Promise<MinioClient>`. So that the credentials will be correctly updated if the token had expired.

Bobgy replied:
I don't know enough background. Is minio client completely stateless?
For example, when there are two requests coming, each request uses a different minio client. Could this be a problem?

Feel free to go ahead if it's not an issue

eterna2 replied:
Yup. It will not be an issue. 

The client actually just constructs the authentication header as well as the corresponding http request based on the configs.

Bobgy replied:
Cool. Sounds good to me

eterna2 commented on frontend/server/server.ts:
```diff
@@ -199,48 +221,26 @@ const artifactsHandler = async (req, res) => {
-      });
+    case 's3':
+      try {
+        const stream = await getObjectStream({bucket, key, client: await s3ClientPromise()});
```
creates the client on-demand to ensure AWS IAM token is updated.

Bobgy commented:
/lgtm

After you confirm my concern in the comment, I will approve

Bobgy commented:
/test kubeflow-pipeline-sample-test

Bobgy commented:
/approve

@eterna2 thanks for the great contribution!

Bobgy commented:
@IronPan @neuromage can you approve? This also has change in manifests.

Ark-kun commented:
> I am on vacation. Not allowed to touch my computer until I am back next mon.

That's great. Good luck for your vacation!

Ark-kun commented:
> @neuromage can you approve? This also has change in manifests.

@neuromage Can you please take a look.

Ark-kun commented:
@neuromage Can you please take a look at manifests and approve? https://github.com/kubeflow/pipelines/pull/2081/files?file-filters%5B%5D=.yaml

Putting this on hold for couple more days due to the KF 0.7 release. We want to be able to easily add small bug fixes if something goes wrong.
/hold

Bobgy commented:
/area front-end

IronPan commented:
@eterna2 I got a general question - how does the argo pod log got persisted in the artifactory in the first place? where did you specify the env var such as ARGO_ARCHIVE_LOGS? Did you set it in workflow controller?

IronPan commented:
Read closer it seems the log persistence is set in argo output artifact repository api
https://github.com/kubeflow/pipelines/pull/2081/files#diff-35aaf752af0c32d077dee5fffb0eadb4R156

does our DSL expose this for now? @Ark-kun

eterna2 commented:
It is not done by kfp. It is currently set by the cluster admin (i.e. via the Argo manifest). Argo allows u to config a default artifactory, and there is an archive logs flag u can set.

https://github.com/argoproj/argo/blob/master/ARTIFACT_REPO.md#configure-the-default-artifact-repository

How my PR handles this when unable to retrieve from the pod directly, is to 
- look at the corresponding Argo workflow crd status. parse the workflow status and check for the archiveLog flag. This is set by the Argo artifactory config in the manifest or inside the workflow template itself (not supported by current kfp sdk).

- but because the status will not be updated properly if the workflow throws an error, the API will fallback to the manual config set inside the front end server itself and makes some simple assumptions.

IronPan commented:
Argo [artifact configmap](https://github.com/kubeflow/pipelines/blob/master/manifests/kustomize/base/argo/workflow-controller-configmap.yaml) is implementation detail that we are not intend to expose, either to cluster user or admin. 

@Ark-kun thoughts?

Ark-kun commented:
I think it's important for cluster administrators to control the artifact storage location.
If we do not want to expose the implementation detail, we should probably provide an official way for configuring the storage.

P.S. I consider archiving the logs to be beneficial. We should enable it by default. Otherwise the logs are lost in non-GCP environments.

IronPan commented:
/hold cancel

mattiasarro commented:
Can anyone give a bit more details on how to enable this? From a master branch installation of Kubeflow, if I set the below env vars on `ml-pipeline-ui`:

```
  ARGO_ARCHIVE_LOGS = "true",
  ARGO_ARCHIVE_ARTIFACTORY = 'minio',
  ARGO_ARCHIVE_BUCKETNAME = 'mlpipeline',
  ARGO_ARCHIVE_PREFIX = 'logs',
```

...and `data.config.artifactRepository.archiveLogs: true` for `workflow-controller-configmap`, I do get logs in Minio `mlpipeline/artifacts/*/*/main.log`, but ml-pipeline-ui does not fetch logs from Minio if a node has been removed.

Is the above setup supposed to work? And is thyhere a way of enabling this for e.g. KF 0.7?

eterna2 commented:
@mattiasarro 

Hmmm. It works on my cluster thou - but I am on AWS eks with my custom manifest rather than kf's manifest.


Can u check what version of pipeline is deployed? And what is the returned error msg?

U might need to update the pipeline-ui service account to have access to k8s secrets and Argo workflow crd.

Because the under the hood, it will first try with k8s API to get the pod logs, followed by getting it from the Argo workflow crd status (which will tell u where the artifacts are stored, it will load the secrets and try to retrieve the logs). And if that fails, it will just try to read the logs based on the env var u set.

I have set it to return diff error msg depending on where it failed.

eterna2 commented:
I just checked. Kf 0.7 uses 0.1.31.

This PR are only merged in for versions > 0.1.34.

U can try using 0.1.34 or 0.1.35. u shld also update the pipeline-ui service account to have access to k8s secrets and Argo workflow crd to.

U can look at the manifest changes in this PR. This is not updated for the main kubeflow manifest repo.

mattiasarro commented:
Thanks @eterna2! 0.1.34 doesn't have the change but 0.1.35 works. I used your [kubeflow-aws](https://github.com/e2fyi/kubeflow-aws) as a basis and created a non-AWS-specific version (since I'm using Minio not S3): https://github.com/mattiasarro/kubeflow-0-7-argo-minio-logs

Ark-kun commented:
@eterna2 Is this feature supposed to work for non-AWS installations (e.g. GCP) as well? There are some reports that it might not be the case.

eterna2 commented:
Yes it should work with both minio and s3.

But there are a few configuration they need to set. By default it is off.

What it does under the hood is just getting the workflow status and retrieving the log artifacts provided in the status.

There are a few ways that it will fail. When workflow error out, sometimes the status will be incomplete - no artifact info. Then in this case, it will not work.

And if secret is required to retrieve the artifacts, the UI sa must have permission to retrieve these secrets.

I don't think I have updated the main kubeflow manifest for this feature.

eterna2 commented:
Is there a good place for us to document some of the configuration guide?
